File: ros2_ws/src/s4_cpp_apps/src/publisher.cpp
#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "s4_custom_interface/msg/hardware_status.hpp"
#include "s4_custom_interface/msg/sphere.hpp"

using namespace std::chrono_literals;

class CppPublisher : public rclcpp::Node {
    public:
        CppPublisher():
            Node("cpp_publisher_node") {
                RCLCPP_INFO(this->get_logger(), "C++ Publisher node has been started");
                publisher_ = this->create_publisher<s4_custom_interface::msg::HardwareStatus>("cpp_hs_topic", 10);
                publisher2_ = this->create_publisher<s4_custom_interface::msg::Sphere>("cpp_sphere_topic", 10);
                timer_ = this->create_wall_timer(500ms, std::bind(&CppPublisher::timer_callback, this));
            }

    private:
        void timer_callback() {
            // HardwareStatus message
            auto message = s4_custom_interface::msg::HardwareStatus();
            message.temperature = 45;
            message.are_motors_up = true;
            message.debug_message = "Eduardo's C++ system runs smoothly";
            message.the_time = this->get_clock()->now();

            RCLCPP_INFO(this->get_logger(), "Publishing HardwareStatus: Temperature: %ld, Motors Up: %s, Debug Message: %s, Time: %d.%09d",
                        message.temperature,
                        message.are_motors_up ? "True" : "False",
                        message.debug_message.c_str(),
                        message.the_time.sec,
                        message.the_time.nanosec);
            
            publisher_->publish(message);

            // Sphere message
            auto message2 = s4_custom_interface::msg::Sphere();
            message2.center.x = 1.0;
            message2.center.y = 2.0;
            message2.center.z = 3.0;
            message2.radius = 4.0;

            RCLCPP_INFO(this->get_logger(), "Publishing Sphere: x: %f, y: %f, z: %f, Radius: %f",
                        message2.center.x,
                        message2.center.y,
                        message2.center.z,
                        message2.radius);
            
            publisher2_->publish(message2);
        }

        rclcpp::TimerBase::SharedPtr timer_;
        rclcpp::Publisher<s4_custom_interface::msg::HardwareStatus>::SharedPtr publisher_;
        rclcpp::Publisher<s4_custom_interface::msg::Sphere>::SharedPtr publisher2_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    std::shared_ptr<rclcpp::Node> node = std::make_shared<CppPublisher>();
    rclcpp::spin(node);
    node.reset();
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/s4_py_apps/s4_py_apps/publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from s4_custom_interface.msg import HardwareStatus
from s4_custom_interface.msg import Sphere

class PyPublisher(Node):
    def __init__(self):
        super().__init__('py_publisher_node')
        self.get_logger().info("Python Publisher node has been started")
        self.publisher_ = self.create_publisher(HardwareStatus, 'py_hs_topic', 10)
        self.publisher2_ = self.create_publisher(Sphere, 'py_sphere_topic', 10)
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)

    def timer_callback(self):
        # HardwareStatus message
        msg = HardwareStatus()
        msg.temperature = 45
        msg.are_motors_up = True
        msg.debug_message = "Eduardo's Python system running smoothly"
        msg.the_time = self.get_clock().now().to_msg()

        self.get_logger().info("Publishing HardwareStatus: Temperature: %d, Motors Up: %s, Debug Message: %s, Time: %d.%09d" % (
                    msg.temperature,
                    msg.are_motors_up,
                    msg.debug_message,
                    msg.the_time.sec,
                    msg.the_time.nanosec))

        self.publisher_.publish(msg)

        # Sphere message
        msg2 = Sphere()
        msg2.center.x = 1.0
        msg2.center.y = 2.0
        msg2.center.z = 3.0
        msg2.radius = 4.0

        self.get_logger().info("Publishing Sphere: x: %f, y: %f, z: %f, Radius: %f" % (
                    msg2.center.x,
                    msg2.center.y,
                    msg2.center.z,
                    msg2.radius))

        self.publisher2_.publish(msg2)

def main(args=None):
    rclpy.init(args=args)
    node = PyPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()